### PR TITLE
Upgrade azure-eventhub to input v2 API

### PR DESCRIPTION
**Pull request by zmoog**
## Proposed commit message

Upgrade azure-eventhub to input v2 API.

The input v2 API provides better health status reporting and control. This is a preliminary step in introducing support for the modern Azure Event Hub SDK for Go. 

## Notes for reviewers

This PR aims to update the azure-eventhub input to the input v2 API without introducing behavior change. The PR is a stepping stone to https://github.com/elastic/beats/issues/33815 (the PR is coming)

Here is a list of changes to keep in mind during the review.

The PR:

- Renames the `azureInput` to `eventHubInputV1` because I want to continue offering the legacy SDK while we introduce an `eventHubInputV2` with the modern event hub SDK in the next PR. The `eventHubInputManager` will select the input version depending on the configuration.
- Switches the pipeline publishing process from Outlet (input v1) to PipelineClient (input v2).
- Starts the EPH and stops it when the input context is "done".
- Rename `parseMultipleMessages()` to `unpackRecords()` because seems to better represent the behavior. 

### Keep existing behavior: Beats pipeline acks

It's important to note that the original input version did not handle publishing ACKs to the Beats pipeline. This was primarily because (AFAIK) the input API v1 did not provide this functionality. 

With the transition to the input API v2, we have acks management. However, the legacy event hub SDK internally updates the checkpoint info after a successful handler call and does not seem to offer hooks to plug proper ACK management.

Since the new modern SDK offers better checkpoint management, in this PR, we keep the current behavior intact, and we'll implement ACKs in the event hub input v2.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Test reload configuration 

## Related issues

- Closes https://github.com/elastic/beats/issues/36377
- Relates https://github.com/elastic/beats/issues/33815

### CODE REVIEW

File: x-pack/filebeat/input/azureeventhub/v1_input_test.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build !aix
+
+package azureeventhub
```
Comment (belimawr) [inline x-pack/filebeat/input/azureeventhub/v1_input_test.go:7]: Empty file? Did something get lost when copying/moving the code or is this file not need any more?
```suggestion
```

**Review by belimawr — COMMENTED** [review]
I took a quick look at the InputV2 API part of the PR and it looks good. I just have a question about an empty file.

File: x-pack/filebeat/input/azureeventhub/v1_input_test.go (hunk 1 above)
Comment (zmoog) [reply x-pack/filebeat/input/azureeventhub/v1_input_test.go:7]: I considered moving some of the tests here but then changed my mind. Thanks for pointing this out. I'm removing it!

File: x-pack/filebeat/input/default-inputs/inputs_other.go [hunk 2]
```diff
@@ -31,6 +32,7 @@ import (
 func xpackInputs(info beat.Info, log *logp.Logger, store beater.StateStore) []v2.Plugin {
 	return []v2.Plugin{
 		azureblobstorage.Plugin(log, store),
+		azureeventhub.Plugin(log),
```
Comment (constanca-m) [inline x-pack/filebeat/input/default-inputs/inputs_other.go:35]: I am a bit confused, what's this for?

File: x-pack/filebeat/input/azureeventhub/eph.go [hunk 3]
```diff
@@ -1,114 +0,0 @@
...
-	azure.PublicCloud.ResourceManagerEndpoint:       azure.PublicCloud,
-	azure.USGovernmentCloud.ResourceManagerEndpoint: azure.USGovernmentCloud,
-}
-
-// runWithEPH will consume ingested events using the Event Processor Host (EPH).
-//
-// To learn more, check the following resources:
-// - https://github.com/Azure/azure-event-hubs-go#event-processor-host
-// - https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-event-processor-host
-func (a *azureInput) runWithEPH() error {
```
Comment (constanca-m) [inline x-pack/filebeat/input/azureeventhub/eph.go:34]: Mainly curious, is there any reason why you moved this code to `v1_input.go`? Is it just related to the legacy input and it's preparation for adding the modern one?

File: x-pack/filebeat/input/default-inputs/inputs_other.go (hunk 2 above)
Comment (zmoog) [reply x-pack/filebeat/input/default-inputs/inputs_other.go:35]: This registers `azureeventhub` (using the input plugin v2 API) and makes it available to Filebeat. Previous, the registration was happening using the old API:

https://github.com/elastic/beats/blob/db9406b50514b8896f08e7bbb176f09c07650566/x-pack/filebeat/input/azureeventhub/input.go#L35-L40

Without registration, the plugin is unknown to Filebeat.

**Comment by zmoog** [comment]
> I cannot comment the lines you did not change, but is there reason to leave this?
> 
> https://github.com/zmoog/beats/blob/cc1d24512dad1a0d8f09055c12cb6e4d49c18ced/x-pack/filebeat/input/azureeventhub/metrics.go#L34
> 
> And again in the test:
> 
> https://github.com/zmoog/beats/blob/cc1d24512dad1a0d8f09055c12cb6e4d49c18ced/x-pack/filebeat/input/azureeventhub/metrics_test.go#L146
> 
> Is it even possible this condition to be true at all?

Your question makes me think I don't fully understand how `metrics.NewUniformSample(1024)` works! 😄 

Can you elaborate on this? Thanks!

**Comment by constanca-m** [comment]
> Your question makes me think I don't fully understand how metrics.NewUniformSample(1024) works! 😄

Indeed, I went to have a look and I think I was confused when I wrote that comment.

I was reading the unit tests to understand the code at first and the `TODO` there got me a bit confused. From `input.processEvents(&ev)` I don't know if the condition can ever be untrue, but this is minor anyway.

File: x-pack/filebeat/input/azureeventhub/v1_input.go [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+
+	return nil
+}
+
+// setup initializes the input components.
+//
+// The main components are:
+// 1. Azure Storage Leaser / Checkpointer
+// 2. Event Processor Host
+// 3. Message handler
```
Comment (constanca-m) [inline x-pack/filebeat/input/azureeventhub/v1_input.go:103]: The comments in the whole PR, especially in this function, are very helpful :+1:

**Review by constanca-m — APPROVED** [review]

File: x-pack/filebeat/input/azureeventhub/input_test.go [hunk 5]
```diff
@@ -119,23 +137,24 @@ func TestParseMultipleRecords(t *testing.T) {
...
-func TestNewInputDone(t *testing.T) {
-	config := mapstr.M{
-		"connection_string":   "Endpoint=sb://something",
-		"eventhub":            "insights-operational-logs",
-		"storage_account":     "someaccount",
-		"storage_account_key": "secret",
-	}
-	inputtest.AssertNotStartedInputCanBeDone(t, NewInput, &config)
-}
+//func TestNewInputDone(t *testing.T) {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/azureeventhub/input_test.go:148]: Should we just remove this test?

**Review by kaiyan-sheng — APPROVED** [review]

**Review by faec — APPROVED** [review]
Looks good!

**Review by cmacknz — APPROVED** [review]
Approving the go.mod changes

**Pull request merged**